product_variant_default_code/models/product.py

```diff
@@ -0,0 +1,206 @@
+# -*- coding: utf-8 -*-
+# Copyright 2014 AvancOSC - Alfredo de la Fuente
+# Copyright 2014 Tecnativa - Pedro M. Baeza
+# Copyright 2014 Shine IT - Tony Gu
+# Copyright 2017 Tecnativa - David Vidal
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+import re
+from string import Template
+from collections import defaultdict
+
+
+class ReferenceMask(Template):
+    pattern = r'''\[(?:
+                    (?P<escaped>\[) |
+                    (?P<named>[^\]]+?)\] |
+                    (?P<braced>[^\]]+?)\] |
+                    (?P<invalid>)
+                    )'''
+
+
+def extract_token(s):
+    pattern = re.compile(r'\[([^\]]+?)\]')
+    return set(pattern.findall(s))
+
+
+def sanitize_reference_mask(product, mask):
+    tokens = extract_token(mask)
+    attribute_names = set()
+    for line in product.attribute_line_ids:
+        attribute_names.add(line.attribute_id.name)
+    if not tokens.issubset(attribute_names):
+        raise UserError(_('Found unrecognized attribute name in "Variant '
+                          'Reference Mask"'))
+
+
+def get_rendered_default_code(product, mask):
+    product_attrs = defaultdict(str)
+    reference_mask = ReferenceMask(mask)
+    for value in product.attribute_value_ids:
+        if value.attribute_id.code:
+            product_attrs[value.attribute_id.name] += value.attribute_id.code
+        if value.code:
+            product_attrs[value.attribute_id.name] += value.code
+    all_attrs = extract_token(mask)
+    missing_attrs = all_attrs - set(product_attrs.keys())
+    missing = dict.fromkeys(
+        missing_attrs,
+        product.env['ir.config_parameter'].get_param(
+            'default_reference_missing_placeholder'))
+    product_attrs.update(missing)
+    default_code = reference_mask.safe_substitute(product_attrs)
+    return default_code
+
+
+def render_default_code(product, mask):
+    product.default_code = get_rendered_default_code(product, mask)
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    prefix_code = fields.Char(
+        string='Prefix code',
```

Comment:
@sebastienbeau could you confirm we really need prefix_code ? because reference mask can be used for that.
and it allow to be better in usability to have not this field.